### PR TITLE
[mcp] Add MCP Server for PMD - enables AI assistants to run static code analysis

### PR DESCRIPTION
**Pull request by roeiba**
## Describe the PR

This PR introduces an **MCP (Model Context Protocol) server** for PMD, enabling AI assistants like Claude, Cursor, Windsurf, and VS Code Copilot to run PMD static code analysis directly.

### Features

- **`pmd_check`** - Run PMD static analysis on source files
- **`pmd_cpd`** - Detect copy-paste (duplicated) code  
- **`pmd_list_languages`** - List supported programming languages
- **`pmd_list_rulesets`** - List available rulesets for analysis

### How it works

The MCP server uses **stdio transport** (like Playwright MCP) for local execution. It wraps the PMD CLI and formats output for LLM consumption.

```
┌─────────────────┐     stdio      ┌─────────────────┐     spawn      ┌─────────────┐
│   AI Client     │ ◄────────────► │   MCP Server    │ ◄────────────► │  PMD CLI    │
│ (Claude/Cursor) │   JSON-RPC     │  (TypeScript)   │   subprocess   │  (Java)     │
└─────────────────┘                └─────────────────┘                └─────────────┘
```

### Usage

Anyone can use it now via npm (published as `@springsoftware/pmd-mcp`):

```json
{
  "mcpServers": {
    "pmd": {
      "command": "npx",
      "args": ["@springsoftware/pmd-mcp"]
    }
  }
}
```

### What's included

- `mcp-pmd/` - New TypeScript project with MCP server implementation
- **41 unit and integration tests** covering all tools and edge cases
- GitHub Actions for CI (build + test with/without PMD) and npm publishing
- Comprehensive documentation (README, MCP_DOCS, ROADMAP)

### Future considerations

If PMD team wants to officially adopt this, the package could be renamed to `@pmd/mcp` and published under the PMD npm organization.

## Related issues

- This is a new feature, no related issues

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (MCP server is TypeScript with separate `npm run build` and `npm test`)
- [x] Added (in-code) documentation (README, MCP_DOCS, ROADMAP)

## Links

- **npm package**: https://www.npmjs.com/package/@springsoftware/pmd-mcp
- **MCP Specification**: https://modelcontextprotocol.io/
- **Playwright MCP (reference)**: https://github.com/microsoft/playwright-mcp

### CODE REVIEW

**Comment by levi-law** [comment]
## Summary

This PR adds an MCP (Model Context Protocol) server for PMD, enabling AI assistants like Claude, Cursor, and Windsurf to run PMD static code analysis directly through natural language interactions.

## What is MCP?

The [Model Context Protocol](https://modelcontextprotocol.io/) is an open standard that allows AI assistants to interact with external tools and data sources. By implementing an MCP server for PMD, we enable seamless integration between AI coding assistants and PMD's powerful static analysis capabilities.

## Features

The MCP server exposes four tools:

| Tool | Description |
|------|-------------|
| `pmd_check` | Run PMD static analysis on source files |
| `pmd_cpd` | Run copy-paste detection (CPD) |
| `pmd_list_languages` | List supported languages for analysis and CPD |
| `pmd_list_rulesets` | List available rulesets, optionally filtered by language |

## Installation & Configuration

### npm Package

```bash
npx @springsoftware/pmd-mcp
```

### MCP Client Configuration

Add to your MCP client (Claude Code, Windsurf, Cursor, Claude Desktop):

```json
{
  "mcpServers": {
    "pmd": {
      "command": "npx",
      "args": ["@springsoftware/pmd-mcp"]
    }
  }
}
```

## Example Usage

### List Supported Languages

**Request:**
```json
{ "tool": "pmd_list_languages" }
```

**Response:**
```
## PMD Supported Languages

### Static Analysis (pmd_check)
- apex, ecmascript, html, java, jsp, kotlin, modelica, plsql, 
  pom, scala, swift, velocity, visualforce, wsdl, xml, xsl

### Copy-Paste Detection (pmd_cpd)
- apex, cpp, cs, dart, ecmascript, go, groovy, java, kotlin, 
  lua, perl, php, python, ruby, rust, scala, swift, typescript, xml
```

### List Rulesets

**Request:**
```json
{ "tool": "pmd_list_rulesets", "arguments": { "language": "java" } }
```

**Response:**
```
## PMD Rulesets

### java Rulesets
- rulesets/java/quickstart.xml - Quick start ruleset with essential rules
- category/java/bestpractices.xml - Best practices and coding standards
- category/java/codestyle.xml - Code style and formatting rules
- category/java/design.xml - Design and architecture rules
- category/java/errorprone.xml - Error-prone code patterns
- category/java/multithreading.xml - Multithreading issues
- category/java/performance.xml - Performance optimization rules
- category/java/security.xml - Security vulnerability detection
```

### Run Static Analysis

**Request:**
```json
{
  "tool": "pmd_check",
  "arguments": {
    "path": "/project/src",
    "rulesets": ["rulesets/java/quickstart.xml"]
  }
}
```

**Response:**
```
## PMD Analysis Results

**Files Analyzed**: 2
**Violations Found**: 7

### 🟡 Medium (Priority 3) (6)

- `/project/src/AdminService.java:14:16` **UnusedLocalVariable**: 
  Avoid unused local variables such as 'temp'.
- `/project/src/AdminService.java:20:9` **ForLoopCanBeForeach**: 
  This for loop can be replaced by a foreach loop
- `/project/src/AdminService.java:34:11` **EmptyCatchBlock**: 
  Avoid empty catch blocks
- `/project/src/UserService.java:15:16` **UnusedLocalVariable**: 
  Avoid unused local variables such as 'temp'.
- `/project/src/UserService.java:20:9` **ForLoopCanBeForeach**: 
  This for loop can be replaced by a foreach loop
- `/project/src/UserService.java:33:11` **EmptyCatchBlock**: 
  Avoid empty catch blocks

### 🔵 Low (Priority 4) (1)

- `/project/src/UserService.java:5:1` **UnnecessaryImport**: 
  Unused import 'java.util.HashMap'
```

### Copy-Paste Detection

**Request:**
```json
{
  "tool": "pmd_cpd",
  "arguments": {
    "path": "/project/src",
    "language": "java",
    "minimum_tokens": 25
  }
}
```

**Response:**
```
## CPD (Copy-Paste Detection) Results

**Language**: java
**Files Analyzed**: 2
**Duplications Found**: 1
**Total Duplicated Lines**: ~16

### Duplications

#### Duplication #1
**16 lines** (61 tokens) duplicated in 2 locations:

- `/project/src/UserService.java:29-44`
- `/project/src/AdminService.java:29-44`
```

## Real-World AI Assistant Interaction

```
User: "Check my Java code for issues and find any duplicated code"

AI Assistant: I'll analyze your Java code using PMD...

[Calls pmd_check with path="/project/src"]
[Calls pmd_cpd with path="/project/src", language="java"]

Found 7 code quality issues:

1. **UnusedLocalVariable** (2 files): Remove unused 'temp' variables
2. **ForLoopCanBeForeach** (2 files): Convert traditional for loops to foreach
3. **EmptyCatchBlock** (2 files): Add proper exception handling
4. **UnnecessaryImport**: Remove unused HashMap import

Also found 1 code duplication:
- The `validateEmail` method is duplicated in UserService.java and AdminService.java
- Consider extracting to a shared utility class

Would you like me to fix these issues?
```

## Technical Details

- **Transport**: stdio (standard input/output)
- **Runtime**: Node.js 18+
- **Dependency**: Requires PMD CLI installed (via Homebrew, or in PATH)
- **Package**: [@springsoftware/pmd-mcp](https://www.npmjs.com/package/@springsoftware/pmd-mcp)
- **MCP Registry**: [io.github.roeiba/pmd](https://registry.modelcontextprotocol.io)

## Files Added

```
mcp-pmd/
├── src/
│   ├── index.ts           # MCP server entry point
│   ├── pmd-executor.ts    # PMD CLI wrapper
│   ├── types.ts           # TypeScript types
│   └── tools/
│       ├── pmd-check.ts   # pmd_check tool
│       ├── pmd-cpd.ts     # pmd_cpd tool
│       └── pmd-list.ts    # pmd_list_* tools
├── tests/                 # Unit and integration tests
├── cli.js                 # CLI entry point
├── package.json
├── tsconfig.json
└── README.md
```

## Testing

```bash
cd mcp-pmd
npm install
npm test
```

## Related Links

- [Model Context Protocol](https://modelcontextprotocol.io/)
- [MCP Registry Entry](https://registry.modelcontextprotocol.io)
- [npm Package](https://www.npmjs.com/package/@springsoftware/pmd-mcp)

**Comment by zbynek** [comment]
The idea of having MCP  wrapper for PMD is interesting, but maybe it's better to have it in a separate repository since it's neither using the same language/tooling or needs to have the same release cycle (?)

Also generally you don't want to put dist folder into a git repo.

File: mcp-pmd/src/tools/pmd-check.ts [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+import { pmdExecutor } from '../pmd-executor.js';
+import type { PmdResult, PmdViolation } from '../types.js';
+
+// Input schema for pmd_check tool
+export const pmdCheckSchema = z.object({
+  path: z.string().describe('Path to source file or directory to analyze'),
+  rulesets: z
+    .array(z.string())
+    .optional()
+    .describe('Rulesets to apply (e.g., ["rulesets/java/quickstart.xml"]). Defaults to quickstart.'),
```
Comment (adangel) [inline mcp-pmd/src/tools/pmd-check.ts:17]: Note, we only have a quickstart ruleset for java and apex. So, we probably can't make this parameter optional.

File: mcp-pmd/src/tools/pmd-check.ts [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+ * 
+ * Runs PMD static code analysis on source files.
+ */
+
+import { z } from 'zod';
+import { pmdExecutor } from '../pmd-executor.js';
+import type { PmdResult, PmdViolation } from '../types.js';
+
+// Input schema for pmd_check tool
+export const pmdCheckSchema = z.object({
```
Comment (adangel) [inline mcp-pmd/src/tools/pmd-check.ts:12]: In order to properly analyze java code, you also need to provide the "aux classpath" - see https://docs.pmd-code.org/latest/pmd_languages_java.html#providing-the-auxiliary-classpath

File: mcp-pmd/src/tools/pmd-list.ts [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+  return lines.join('\n');
+}
+
+/**
+ * Execute list rulesets tool
+ */
+export async function executeListRulesets(language?: string): Promise<string> {
+  // Common rulesets by language
+  const rulesetsByLanguage: Record<string, Array<{ name: string; description: string }>> = {
```
Comment (adangel) [inline mcp-pmd/src/tools/pmd-list.ts:43]: Looks like implementing #1335 would be useful here as well in the future

File: mcp-pmd/src/pmd-executor.ts [hunk 4]
```diff
@@ -0,0 +1,507 @@
...
+
+  /**
+   * Parse CPD XML output
+   */
+  private parseCpdXmlOutput(output: string): CpdResult {
+    const duplications: CpdDuplication[] = [];
+    let filesAnalyzed = 0;
+
+    try {
+      // Simple XML parsing for CPD output
```
Comment (adangel) [inline mcp-pmd/src/pmd-executor.ts:382]: Maybe a json-based format for CPD would be useful here in the future...

File: mcp-pmd/src/pmd-executor.ts [hunk 5]
```diff
@@ -0,0 +1,507 @@
...
+        processingErrors: ['Failed to parse CPD output'],
+      };
+    }
+  }
+
+  /**
+   * List supported languages
+   */
+  async listLanguages(): Promise<string[]> {
+    // PMD supported languages (from documentation)
```
Comment (adangel) [inline mcp-pmd/src/pmd-executor.ts:441]: You could parse this from `pmd check --help` or `pmd cpd --help`.

**Comment by roeiba** [comment]
Thank you @zbynek and @adangel for the feedback and for creating the dedicated repository!

I've migrated the MCP server code to the new pmd-mcp repository and opened a PR there:
👉 **https://github.com/pmd/pmd-mcp/pull/1**

Changes made based on your feedback:
- ✅ Moved code to the root folder (removed \`mcp-pmd/\` prefix)
- ✅ Removed \`dist/\` from git tracking (proper .gitignore)
- ✅ Updated package.json URLs to point to pmd/pmd-mcp

I'll address the technical feedback (aux classpath, ruleset requirements, parsing languages from --help) in follow-up commits to the new PR.

This PR can be closed. Thanks again!

**Comment by adangel** [comment]
Awesome! Thank you!
